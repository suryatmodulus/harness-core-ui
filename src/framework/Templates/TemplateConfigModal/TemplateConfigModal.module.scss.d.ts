/* eslint-disable */
// this is an auto-generated file
declare const styles: {
  readonly basicDetails: string
  readonly nameIdDescriptionTags: string
  readonly comment: string
  readonly closeIcon: string
}
export default styles
