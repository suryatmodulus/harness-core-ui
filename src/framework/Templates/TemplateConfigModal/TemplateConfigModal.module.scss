.basicDetails {
  padding: var(--spacing-10) var(--spacing-11) !important;
  border-radius: 6px;

  .nameIdDescriptionTags {
    margin-bottom: 0px !important;
    > div {
      margin-bottom: var(--spacing-medium) !important;
    }
  }

  .comment {
    max-height: 240px !important;
  }
}

.closeIcon {
  position: absolute;
  top: var(--spacing-6);
  right: var(--spacing-4);
}
