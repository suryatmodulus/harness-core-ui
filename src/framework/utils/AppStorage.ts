/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck
/* eslint-disable @typescript-eslint/explicit-function-return-type */
// ignoring ts and eslint issues because this is a legacy file

////////////////////////////////////////////////////////////////////////
// THIS FILE IS A LEGACY - DO NOT USE IT DIRECTLY OR EXTEND ITS USAGE //
////////////////////////////////////////////////////////////////////////

import * as SecureStorage from 'secure-web-storage'

const SecureStorageConstructor = SecureStorage.default
const keyStr = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/='

const buildLoginUrlFrom401Response = () => {
  const { href } = window.location
  const prefix = window.HARNESS_ENABLE_NG_AUTH_UI ? '/auth/#/signin?returnUrl=' : '#/login?returnUrl='
  return href.includes(prefix) ? href : prefix + encodeURIComponent(href)
}

function encode64(input) {
  input = escape(input)
  let output = ''
  let chr1,
    chr2,
    chr3 = ''
  let enc1,
    enc2,
    enc3,
    enc4 = ''
  let i = 0

  do {
    chr1 = input.charCodeAt(i++)
    chr2 = input.charCodeAt(i++)
    chr3 = input.charCodeAt(i++)

    enc1 = chr1 >> 2
    enc2 = ((chr1 & 3) << 4) | (chr2 >> 4)
    enc3 = ((chr2 & 15) << 2) | (chr3 >> 6)
    enc4 = chr3 & 63

    if (isNaN(chr2)) {
      enc3 = enc4 = 64
    } else if (isNaN(chr3)) {
      enc4 = 64
    }

    output = output + keyStr.charAt(enc1) + keyStr.charAt(enc2) + keyStr.charAt(enc3) + keyStr.charAt(enc4)
    chr1 = chr2 = chr3 = ''
    enc1 = enc2 = enc3 = enc4 = ''
  } while (i < input.length)

  return output
}

function decode64(input) {
  let output = ''
  let chr1,
    chr2,
    chr3 = ''
  let enc1,
    enc2,
    enc3,
    enc4 = ''
  let i = 0

  // remove all characters that are not A-Z, a-z, 0-9, +, /, or =
  const base64test = /[^A-Za-z0-9+/=]/g
  input = input.replace(base64test, '')

  do {
    enc1 = keyStr.indexOf(input.charAt(i++))
    enc2 = keyStr.indexOf(input.charAt(i++))
    enc3 = keyStr.indexOf(input.charAt(i++))
    enc4 = keyStr.indexOf(input.charAt(i++))

    chr1 = (enc1 << 2) | (enc2 >> 4)
    chr2 = ((enc2 & 15) << 4) | (enc3 >> 2)
    chr3 = ((enc3 & 3) << 6) | enc4

    output = output + String.fromCharCode(chr1)

    if (enc3 !== 64) {
      output = output + String.fromCharCode(chr2)
    }
    if (enc4 !== 64) {
      output = output + String.fromCharCode(chr3)
    }

    chr1 = chr2 = chr3 = ''
    enc1 = enc2 = enc3 = enc4 = ''
  } while (i < input.length)

  return unescape(output)
}

const secureStorage = new SecureStorageConstructor(localStorage, {
  hash: function hash(key) {
    return key
  },
  encrypt: function encrypt(data) {
    return encode64(data)
  },
  decrypt: function decrypt(data) {
    return decode64(data)
  }
})

export default class AppStorage {
  static set(name, value) {
    if (value && name !== 'DataStore_apps') {
      secureStorage.setItem(name, value)
    }
    if (value && name === 'DataStore_apps') {
      localStorage.setItem(name, value)
    }
  }

  static get(name) {
    try {
      return secureStorage.getItem(name)
    } catch (event) {
      window.location = buildLoginUrlFrom401Response()
    }
  }

  static getAll() {
    const _obj = {}

    for (let i = 0, len = secureStorage.length; i < len; ++i) {
      const key = secureStorage.key(i)
      _obj.key = AppStorage.get(key)
    }

    return _obj
  }

  static has(name) {
    return AppStorage.get(name) ? true : false
  }

  static remove(name) {
    if (name !== 'DataStore_apps') {
      secureStorage.removeItem(name)
    } else {
      localStorage.removeItem(name)
    }
  }

  static cleanupOnAccountSwitch() {
    // clear AppStorage, only keep important fields:
    const storage = {
      token: AppStorage.get('token'),
      username: AppStorage.get('username') || '',
      email: AppStorage.get('email') || ''
    }

    AppStorage.clear()

    Object.keys(storage).forEach(key => {
      AppStorage.set(key, storage[key])
    })
  }

  static clear() {
    secureStorage.clear()
    localStorage.clear()
  }

  static decode64 = decode64
}
