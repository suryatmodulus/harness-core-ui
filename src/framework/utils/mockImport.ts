/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

/**
 * Test utility to mock any import. It's better than jest.mock() because you can use
 * mockImport() inside tests.
 *
 * Sample:
 *
 *  mockImport('services/cf', {
 *    useGetFeatureFlag: () => ({
 *      data: mockFeatureFlag,
 *      loading: undefined,
 *      error: undefined,
 *      refetch: jest.fn()
 *     })
 *  })
 *
 * Mock an `export default`:
 *
 *  mockImport('@cf/components/FlagActivation/FlagActivation', {
 *     // FlagActivation is exported as `export default`
 *     default: function FlagActivation() {
 *       return <div>FlagActivation</div>
 *    }
 *  })
 *
 * @param moduleName
 * @param implementation
 */
export default function mockImport(moduleName: string, implementation: Record<string, any>): void {
  // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires, global-require
  const module = require(moduleName)

  Object.keys(implementation).forEach(key => {
    module[key] = implementation?.[key]
  })
}
