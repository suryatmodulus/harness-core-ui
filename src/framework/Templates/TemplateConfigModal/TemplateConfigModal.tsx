import React, { Dispatch, useState, SetStateAction, useContext } from 'react'
import * as Yup from 'yup'
import { isEmpty, omit } from 'lodash-es'
import type { FormikProps } from 'formik'
import {
  Formik,
  Layout,
  FormikForm,
  FormInput,
  Text,
  Color,
  Button,
  ButtonVariation,
  Container
} from '@wings-software/uicore'
import { useStrings } from 'framework/strings'
import { NameIdDescriptionTags } from '@common/components/NameIdDescriptionTags/NameIdDescriptionTags'
import RbacButton from '@rbac/components/Button/Button'
import { PermissionIdentifier } from '@rbac/interfaces/PermissionIdentifier'
import { ResourceType } from '@rbac/interfaces/ResourceType'
import type { EntityGitDetails, NGTemplateInfoConfig } from 'services/template-ng'
import { TemplatePreview } from '@templates-library/components/TemplatePreview/TemplatePreview'
import { TemplateContext } from '@templates-library/components/TemplateStudio/TemplateContext/TemplateContext'
import { PageSpinner } from '@common/components'
import type { UseSaveSuccessResponse } from '@common/modals/SaveToGitDialog/useSaveToGitDialog'
import GitContextForm, { IGitContextFormProps } from '@common/components/GitContextForm/GitContextForm'
import { useAppStore } from 'framework/AppStore/AppStoreContext'
import { GitSyncStoreProvider } from 'framework/GitRepoStore/GitSyncStoreContext'
import { IdentifierSchema, NameSchema } from '@common/utils/Validation'
import { regexVersionLabel } from '@common/utils/StringUtils'
import type { Error } from 'services/cd-ng'
import { DefaultNewTemplateId, DefaultNewVersionLabel } from '../templates'
import css from './TemplateConfigModal.module.scss'

export enum Fields {
  Name = 'name',
  Identifier = 'identifier',
  Description = 'description',
  Tags = 'tags',
  VersionLabel = 'versionLabel',
  Repo = 'repo',
  Branch = 'branch'
}

export interface PromiseExtraArgs {
  isEdit?: boolean
  updatedGitDetails?: EntityGitDetails
  comment?: string
}

export interface ModalProps {
  title: string
  disabledFields?: Fields[]
  emptyFields?: Fields[]
  shouldGetComment?: boolean
  promise: (values: NGTemplateInfoConfig, extraInfo: PromiseExtraArgs) => Promise<UseSaveSuccessResponse>
  onSuccess?: (values: NGTemplateInfoConfig) => void
  onFailure?: (error: Error) => void
}

export interface TemplateConfigValues extends NGTemplateInfoConfigWithGitDetails {
  comment: string
}

export interface NGTemplateInfoConfigWithGitDetails extends NGTemplateInfoConfig {
  repo: string
  branch: string
}

export interface ConfigModalProps {
  initialValues: NGTemplateInfoConfigWithGitDetails
  onClose: () => void
  modalProps: ModalProps
  showGitFields?: boolean
  gitDetails?: IGitContextFormProps
}

interface BasicDetailsInterface extends ConfigModalProps {
  setPreviewValues: Dispatch<SetStateAction<NGTemplateInfoConfigWithGitDetails>>
}

const MAX_VERSION_LABEL_LENGTH = 63

const BasicTemplateDetails = (props: BasicDetailsInterface): JSX.Element => {
  const { initialValues, setPreviewValues, onClose, modalProps, showGitFields, gitDetails } = props
  const { title, disabledFields = [], shouldGetComment = false, promise, onSuccess, onFailure } = modalProps
  const { getString } = useStrings()
  const [isEdit, setIsEdit] = React.useState<boolean>()
  const { isGitSyncEnabled } = useAppStore()
  const currentTemplateType = initialValues.type
  const formName = `create${currentTemplateType}Template`
  const [loading, setLoading] = React.useState<boolean>()
  const { isReadonly } = useContext(TemplateContext)

  React.useEffect(() => {
    const edit = initialValues.identifier !== DefaultNewTemplateId
    setIsEdit(edit)
  }, [initialValues])

  const onSubmit = React.useCallback((values: TemplateConfigValues) => {
    setLoading(true)
    const formGitDetails =
      values.repo && values.repo.trim().length > 0 ? { repoIdentifier: values.repo, branch: values.branch } : undefined
    const comment = values.comment.trim()
    promise(omit(values, 'repo', 'branch', 'comment'), {
      isEdit: false,
      updatedGitDetails: formGitDetails,
      ...(comment.length > 0 && { comment: values.comment })
    })
      .then(response => {
        setLoading(false)
        if (response && response.status === 'SUCCESS') {
          onSuccess?.(values)
          onClose()
        } else {
          throw response
        }
      })
      .catch(error => {
        setLoading(false)
        onFailure?.(error)
      })
  }, [])

  const getSaveButtonText = React.useCallback((): string => {
    if (isEdit) {
      if (isGitSyncEnabled) {
        return getString('continue')
      }
      return getString('save')
    }
    return getString('start')
  }, [isEdit, isGitSyncEnabled, getString])

  const versionLabelText = getString('templatesLibrary.createNewModal.versionLabel')

  return (
    <Container width={'55%'} className={css.basicDetails} background={Color.FORM_BG} padding={'huge'}>
      {loading && <PageSpinner />}
      <Text
        color={Color.GREY_800}
        font={{ weight: 'bold', size: 'medium' }}
        margin={{ bottom: 'xlarge', left: 0, right: 0 }}
      >
        {title || ''}
      </Text>
      <Formik<TemplateConfigValues>
        initialValues={{ ...initialValues, comment: '' }}
        onSubmit={onSubmit}
        validate={values => setPreviewValues(values)}
        formName={formName}
        enableReinitialize={true}
        validationSchema={Yup.object().shape({
          name: NameSchema({
            requiredErrorMsg: getString('common.validation.fieldIsRequired', {
              name: getString('templatesLibrary.createNewModal.nameError')
            })
          }),
          identifier: IdentifierSchema(),
          versionLabel: Yup.string()
            .trim()
            .required(
              getString('common.validation.fieldIsRequired', {
                name: versionLabelText
              })
            )
            .matches(
              regexVersionLabel,
              getString('common.validation.fieldMustStartWithAlphanumericAndCanNotHaveSpace', {
                name: versionLabelText
              })
            )
            .max(
              MAX_VERSION_LABEL_LENGTH,
              getString('common.validation.fieldCannotbeLongerThanN', {
                name: versionLabelText,
                n: MAX_VERSION_LABEL_LENGTH
              })
            ),
          ...(isGitSyncEnabled && showGitFields
            ? {
                repo: Yup.string().trim().required(getString('common.git.validation.repoRequired')),
                branch: Yup.string().trim().required(getString('common.git.validation.branchRequired'))
              }
            : {})
        })}
      >
        {(formik: FormikProps<TemplateConfigValues>) => {
          return (
            <FormikForm>
              <Layout.Vertical spacing={'huge'}>
                <Container>
                  <Layout.Vertical spacing={'small'}>
                    <Container>
                      <Layout.Vertical>
                        <NameIdDescriptionTags
                          tooltipProps={{ dataTooltipId: formName }}
                          formikProps={formik}
                          identifierProps={{
                            isIdentifierEditable: !disabledFields.includes(Fields.Identifier) && !isReadonly,
                            inputGroupProps: { disabled: disabledFields.includes(Fields.Name) || isReadonly }
                          }}
                          className={css.nameIdDescriptionTags}
                          descriptionProps={{
                            disabled: disabledFields.includes(Fields.Description) || isReadonly
                          }}
                          tagsProps={{
                            disabled: disabledFields.includes(Fields.Tags) || isReadonly
                          }}
                        />
                        <FormInput.Text
                          name="versionLabel"
                          placeholder={getString('templatesLibrary.createNewModal.versionPlaceholder')}
                          label={versionLabelText}
                          disabled={disabledFields.includes(Fields.VersionLabel) || isReadonly}
                        />
                        {shouldGetComment && (
                          <FormInput.TextArea
                            name="comment"
                            label={getString('optionalField', {
                              name: getString('common.commentModal.commentLabel')
                            })}
                            className={css.comment}
                          />
                        )}
                      </Layout.Vertical>
                    </Container>
                    {isGitSyncEnabled && showGitFields && (
                      <GitSyncStoreProvider>
                        <GitContextForm formikProps={formik} gitDetails={gitDetails} />
                      </GitSyncStoreProvider>
                    )}
                  </Layout.Vertical>
                </Container>
                <Container>
                  <Layout.Horizontal spacing="small" flex={{ alignItems: 'flex-end', justifyContent: 'flex-start' }}>
                    <RbacButton
                      text={getSaveButtonText()}
                      type="submit"
                      variation={ButtonVariation.PRIMARY}
                      permission={{
                        permission: PermissionIdentifier.EDIT_TEMPLATE,
                        resource: {
                          resourceType: ResourceType.TEMPLATE
                        }
                      }}
                    />
                    <Button text={getString('cancel')} variation={ButtonVariation.TERTIARY} onClick={onClose} />
                  </Layout.Horizontal>
                </Container>
              </Layout.Vertical>
            </FormikForm>
          )
        }}
      </Formik>
    </Container>
  )
}

export const TemplateConfigModal = (props: ConfigModalProps): JSX.Element => {
  const { initialValues, modalProps, ...rest } = props
  const { emptyFields = [] } = modalProps
  const [previewValues, setPreviewValues] = useState(props.initialValues)
  const [formInitialValues, setFormInitialValues] = React.useState(initialValues)
  const { isGitSyncEnabled } = useAppStore()
  const { getString } = useStrings()

  React.useEffect(() => {
    const newInitialValues = {
      ...initialValues,
      ...((emptyFields.includes(Fields.Name) || isEmpty(initialValues.name)) && {
        name: getString('templatesLibrary.createNewModal.namePlaceholder', { entity: initialValues.type })
      }),
      ...((emptyFields.includes(Fields.Identifier) || initialValues.identifier === DefaultNewTemplateId) && {
        identifier: getString('templatesLibrary.createNewModal.identifierPlaceholder', {
          entity: initialValues.type.toLowerCase()
        })
      }),
      ...((emptyFields.includes(Fields.VersionLabel) || initialValues.versionLabel === DefaultNewVersionLabel) && {
        versionLabel: 'Version1'
      }),
      ...(emptyFields.includes(Fields.Description) && { description: undefined }),
      ...(emptyFields.includes(Fields.Tags) && { tags: undefined }),
      ...(emptyFields.includes(Fields.Repo) && { repo: undefined }),
      ...(emptyFields.includes(Fields.Branch) && { branch: undefined })
    }
    setFormInitialValues(newInitialValues)
  }, [initialValues, JSON.stringify(emptyFields)])

  React.useEffect(() => {
    setPreviewValues(formInitialValues)
  }, [formInitialValues])

  const content = (
    <Layout.Horizontal>
      <BasicTemplateDetails
        initialValues={formInitialValues}
        modalProps={modalProps}
        setPreviewValues={setPreviewValues}
        {...rest}
      />
      <TemplatePreview previewValues={previewValues} />
      <Button
        className={css.closeIcon}
        iconProps={{ size: 24, color: Color.GREY_500 }}
        icon="cross"
        variation={ButtonVariation.ICON}
        onClick={props.onClose}
      />
    </Layout.Horizontal>
  )
  return isGitSyncEnabled ? <GitSyncStoreProvider>{content}</GitSyncStoreProvider> : content
}
