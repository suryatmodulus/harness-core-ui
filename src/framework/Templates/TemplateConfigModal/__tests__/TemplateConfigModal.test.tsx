/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render, act, fireEvent, waitFor } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import { ConfigModalProps, TemplateConfigModal } from '../TemplateConfigModal'

const getProps = (): ConfigModalProps => ({
  onClose: jest.fn(),
  initialValues: {
    name: '',
    type: 'Step',
    identifier: 'My_Step_Template',
    versionLabel: 'v1',
    repo: 'test_repo',
    branch: 'test_branch'
  },
  modalProps: {
    title: 'Some Title',
    promise: Promise.resolve
  },
  showGitFields: false
})

describe('CREATE MODE', () => {
  test('VALIDATIONS', async () => {
    const props = getProps()
    const { container, getByText, queryByText } = render(
      <TestWrapper>
        <TemplateConfigModal {...props} />
      </TestWrapper>
    )
    act(() => {
      fireEvent.change(container.querySelector('input[name="name"]')!, { target: { value: '' } })
    })
    act(() => {
      fireEvent.click(getByText('save'))
    })
    await waitFor(() => expect(queryByText('common.validation.fieldIsRequired')).toBeTruthy())
  })

  test('if form changesupdate the preview card', async () => {
    const props = getProps()
    const { container } = render(
      <TestWrapper>
        <TemplateConfigModal {...props} />
      </TestWrapper>
    )
    act(() => {
      fireEvent.change(container.querySelector('input[name="name"]')!, { target: { value: 'templatename' } })
    })
    expect(container).toMatchSnapshot('changed value should be present in the preview')
  })

  test('if form closes', async () => {
    const props = getProps()
    const { container, getByText } = render(
      <TestWrapper>
        <TemplateConfigModal {...props} />
      </TestWrapper>
    )
    act(() => {
      fireEvent.click(getByText('cancel'))
    })
    await waitFor(() => expect(props.onClose).toBeCalled())

    const crossIcon = container.querySelector('span[icon="cross"]')
    act(() => {
      fireEvent.click(crossIcon!)
    })
    await waitFor(() => expect(props.onClose).toBeCalled())
  })
})
