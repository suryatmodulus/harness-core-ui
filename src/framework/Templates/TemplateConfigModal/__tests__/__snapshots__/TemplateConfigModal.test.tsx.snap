// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CREATE MODE if form changesupdate the preview card: changed value should be present in the preview 1`] = `
<div>
  <div
    class="Layout--horizontal StyledProps--main"
  >
    <div
      class="StyledProps--font StyledProps--main basicDetails StyledProps--background-form StyledProps--padding-huge"
      style="width: 55%;"
    >
      <p
        class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800 StyledProps--font-weight-bold StyledProps--font-size-medium StyledProps--margin-bottom-xlarge"
      >
        Some Title
      </p>
      <div
        class="OverlaySpinner--overlaySpinner"
      >
        <form
          class="FormikForm--main"
        >
          <div>
            <div
              class="Layout--vertical Layout--layout-spacing-huge StyledProps--main"
            >
              <div
                class="StyledProps--font StyledProps--main"
              >
                <div
                  class="Layout--vertical Layout--layout-spacing-small StyledProps--main"
                >
                  <div
                    class="StyledProps--font StyledProps--main"
                  >
                    <div
                      class="Layout--vertical StyledProps--main"
                    >
                      <div
                        class="StyledProps--font StyledProps--main main nameIdDescriptionTags"
                      >
                        <div
                          class="InputWithIdentifier--txtNameContainer TextInput--main"
                        >
                          <div
                            class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main InputWithIdentifier--txtIdContainer"
                          >
                            <div
                              class="Text--withDocsTooltip"
                              data-tooltip-id="createStepTemplate_identifier"
                            >
                              <p
                                class="StyledProps--font StyledProps--main"
                              >
                                ID
                              </p>
                            </div>
                            <p
                              class="StyledProps--font StyledProps--main"
                            >
                              :
                            </p>
                            <span
                              class="bp3-popover-wrapper"
                            >
                              <span
                                class="bp3-popover-target"
                              >
                                <div
                                  class="bp3-editable-text InputWithIdentifier--idValue"
                                  tabindex="0"
                                >
                                  <span
                                    class="bp3-editable-text-content"
                                    style="height: 1px; line-height: 1px; min-width: 0;"
                                  >
                                    templatename
                                  </span>
                                </div>
                              </span>
                            </span>
                            <span
                              class="bp3-icon StyledProps--main"
                              data-icon="Edit"
                              style="vertical-align: middle; cursor: pointer;"
                            >
                              <svg
                                height="12"
                                viewBox="0 0 16 16"
                                width="12"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  clip-rule="evenodd"
                                  d="M11.207.485a.5.5 0 01.354-.146 3.5 3.5 0 013.5 3.5.5.5 0 01-.146.353l-11 11a.5.5 0 01-.354.147h-3a.5.5 0 01-.5-.5v-3a.5.5 0 01.146-.354l11-11zm.553.862L1.061 12.046v2.293h2.293l10.7-10.7a2.5 2.5 0 00-2.294-2.292z"
                                  fill="currentColor"
                                  fill-rule="evenodd"
                                />
                                <path
                                  clip-rule="evenodd"
                                  d="M11.89 2.646a.5.5 0 010 .708L8.353 6.889a.5.5 0 01-.708-.707l3.536-3.536a.5.5 0 01.707 0z"
                                  fill="currentColor"
                                  fill-rule="evenodd"
                                  opacity="0.3"
                                />
                              </svg>
                            </span>
                          </div>
                          <div
                            class="bp3-form-group"
                            data-id="name-0"
                          >
                            <label
                              class="bp3-label"
                              for="name"
                            >
                              <span
                                class="Tooltip--acenter"
                                data-tooltip-id="createStepTemplate_name"
                              >
                                name
                              </span>
                               
                              <span
                                class="bp3-text-muted"
                              />
                            </label>
                            <div
                              class="bp3-form-content"
                            >
                              <div
                                class="bp3-input-group"
                              >
                                <input
                                  autocomplete="off"
                                  class="bp3-input"
                                  name="name"
                                  type="text"
                                  value="templatename"
                                />
                              </div>
                            </div>
                          </div>
                        </div>
                        <div
                          class="StyledProps--font StyledProps--main"
                        >
                          <label
                            class="Label--main bp3-label descriptionLabel"
                            data-tooltip-id="createStepTemplate_description"
                          >
                            optionalField
                            <span
                              class="bp3-icon StyledProps--main editOpen"
                              data-icon="Edit"
                              data-name="edit"
                              data-testid="description-edit"
                            >
                              <svg
                                height="12"
                                viewBox="0 0 16 16"
                                width="12"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  clip-rule="evenodd"
                                  d="M11.207.485a.5.5 0 01.354-.146 3.5 3.5 0 013.5 3.5.5.5 0 01-.146.353l-11 11a.5.5 0 01-.354.147h-3a.5.5 0 01-.5-.5v-3a.5.5 0 01.146-.354l11-11zm.553.862L1.061 12.046v2.293h2.293l10.7-10.7a2.5 2.5 0 00-2.294-2.292z"
                                  fill="currentColor"
                                  fill-rule="evenodd"
                                />
                                <path
                                  clip-rule="evenodd"
                                  d="M11.89 2.646a.5.5 0 010 .708L8.353 6.889a.5.5 0 01-.708-.707l3.536-3.536a.5.5 0 01.707 0z"
                                  fill="currentColor"
                                  fill-rule="evenodd"
                                  opacity="0.3"
                                />
                              </svg>
                            </span>
                          </label>
                        </div>
                        <div
                          class="StyledProps--font StyledProps--main"
                        >
                          <label
                            class="Label--main bp3-label descriptionLabel"
                            data-tooltip-id="createStepTemplate_tags"
                          >
                            optionalField
                            <span
                              class="bp3-icon StyledProps--main editOpen"
                              data-icon="Edit"
                              data-name="edit"
                              data-testid="tags-edit"
                            >
                              <svg
                                height="12"
                                viewBox="0 0 16 16"
                                width="12"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  clip-rule="evenodd"
                                  d="M11.207.485a.5.5 0 01.354-.146 3.5 3.5 0 013.5 3.5.5.5 0 01-.146.353l-11 11a.5.5 0 01-.354.147h-3a.5.5 0 01-.5-.5v-3a.5.5 0 01.146-.354l11-11zm.553.862L1.061 12.046v2.293h2.293l10.7-10.7a2.5 2.5 0 00-2.294-2.292z"
                                  fill="currentColor"
                                  fill-rule="evenodd"
                                />
                                <path
                                  clip-rule="evenodd"
                                  d="M11.89 2.646a.5.5 0 010 .708L8.353 6.889a.5.5 0 01-.708-.707l3.536-3.536a.5.5 0 01.707 0z"
                                  fill="currentColor"
                                  fill-rule="evenodd"
                                  opacity="0.3"
                                />
                              </svg>
                            </span>
                          </label>
                        </div>
                      </div>
                      <div
                        class="bp3-form-group"
                        data-id="versionLabel-1"
                      >
                        <label
                          class="bp3-label"
                          for="versionLabel"
                        >
                          <span
                            class="Tooltip--acenter"
                            data-tooltip-id="createStepTemplate_versionLabel"
                          >
                            templatesLibrary.createNewModal.versionLabel
                          </span>
                           
                          <span
                            class="bp3-text-muted"
                          />
                        </label>
                        <div
                          class="bp3-form-content"
                        >
                          <div
                            class="bp3-input-group"
                          >
                            <input
                              autocomplete="off"
                              class="bp3-input"
                              name="versionLabel"
                              placeholder="templatesLibrary.createNewModal.versionPlaceholder"
                              type="text"
                              value="v1"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="StyledProps--font StyledProps--main"
              >
                <div
                  class="Layout--horizontal Layout--layout-spacing-small StyledProps--main StyledProps--flex StyledProps--flex-alignItems-flex-end StyledProps--flex-justifyContent-flex-start"
                >
                  <button
                    class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--variation Button--variation-primary"
                    type="submit"
                  >
                    <span
                      class="bp3-button-text"
                    >
                      save
                    </span>
                  </button>
                  <button
                    class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--variation Button--variation-tertiary"
                    type="button"
                  >
                    <span
                      class="bp3-button-text"
                    >
                      cancel
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
    <div
      class="Layout--horizontal StyledProps--main preview StyledProps--flex StyledProps--flex-alignItems-center StyledProps--flex-justifyContent-center StyledProps--padding-top-huge StyledProps--padding-bottom-huge"
    >
      <div
        class="StyledProps--font StyledProps--main container"
      >
        <div
          class="bp3-card bp3-elevation-0 Card--card templateCard"
        >
          <div />
          <div
            class="StyledProps--font StyledProps--main StyledProps--margin-right-small"
          >
            <div
              class="Layout--horizontal Layout--layout-spacing-small StyledProps--main StyledProps--margin-bottom-small StyledProps--flex"
            >
              <span
                class="bp3-icon bp3-icon-disable StyledProps--main"
                icon="disable"
              >
                <svg
                  data-icon="disable"
                  height="24"
                  viewBox="0 0 20 20"
                  width="24"
                >
                  <desc>
                    disable
                  </desc>
                  <path
                    d="M10 0C4.48 0 0 4.48 0 10s4.48 10 10 10 10-4.48 10-10S15.52 0 10 0zM2 10c0-4.42 3.58-8 8-8 1.85 0 3.55.63 4.9 1.69L3.69 14.9A7.902 7.902 0 012 10zm8 8c-1.85 0-3.55-.63-4.9-1.69L16.31 5.1A7.902 7.902 0 0118 10c0 4.42-3.58 8-8 8z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </div>
          </div>
          <div
            class="StyledProps--font StyledProps--main"
          >
            <p
              class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--font-weight-semi-bold StyledProps--color StyledProps--color-grey800"
              data-testid="templatename"
              style="--text-line-clamp: 1;"
            >
              templatename
            </p>
            <p
              class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--font-small StyledProps--color StyledProps--color-grey400 StyledProps--margin-top-xsmall"
              style="--text-line-clamp: 1;"
            >
              idLabel
            </p>
          </div>
          <div
            class="StyledProps--font StyledProps--main StyledProps--background-grey100"
            style="height: 1px;"
          />
          <div
            class="StyledProps--font StyledProps--main"
          >
            <span
              class="Tag--main"
            >
              <span
                class="bp3-tag version"
              >
                <span
                  class="bp3-text-overflow-ellipsis bp3-fill"
                >
                  v1
                </span>
              </span>
            </span>
          </div>
          <div
            class="StyledProps--font StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-center StyledProps--padding-top-medium"
          >
            <div
              class="templateColor"
            >
              <div
                class="absolutePos"
              >
                <svg
                  fill="none"
                  height="18"
                  viewBox="0 0 121 18"
                  width="121"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M10.6577 4.36758L0.989019 17.5H120.044L111.076 4.58226C109.301 2.02513 106.385 0.5 103.272 0.5H18.3079C15.288 0.5 12.4481 1.93575 10.6577 4.36758Z"
                    fill="#EADEFF"
                    stroke="#E1D0FF"
                  />
                </svg>
              </div>
              <p
                class="StyledProps--font StyledProps--main text absolutePos"
                style="color: rgb(89, 43, 170);"
              >
                STEP
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <button
      class="bp3-button Button--button StyledProps--font StyledProps--main closeIcon Button--with-current-color Button--iconOnly Button--withLeftIcon Button--variation Button--variation-icon"
      type="button"
    >
      <span
        class="bp3-icon bp3-icon-cross StyledProps--main StyledProps--color StyledProps--color-grey500"
        icon="cross"
      >
        <svg
          data-icon="cross"
          height="24"
          viewBox="0 0 20 20"
          width="24"
        >
          <desc>
            cross
          </desc>
          <path
            d="M11.41 10l4.29-4.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L10 8.59l-4.29-4.3a1.003 1.003 0 00-1.42 1.42L8.59 10 4.3 14.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l4.29-4.3 4.29 4.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L11.41 10z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
